# Decision Tree Regression Project  

## Tesla Model Y Price Prediction | End-to-End ML Project  

This project focuses on predicting the price of Tesla Model Y cars using **Decision Tree Regression**.

It covers the full machine learning workflow, including data preprocessing, exploratory data analysis (EDA), feature selection, and model training.

---

## Dataset  

The dataset used in this project is available on Kaggle: [Dataset Link](https://www.kaggle.com/datasets/aiwithcagri/multiple-linear-regression-tesla-model-y-price)
  

---

## Objective  

The main objective is to build a regression model that can accurately estimate car prices based on key features such as:
- year
- km  
- color 

---

## Workflow  

The project follows these steps:

1. **Data Cleaning**

2. **Exploratory Data Analysis (EDA)**

3. **Feature Engineering** 

4. **Model Building**

5. **Model Evaluation**

---

## Tools & Technologies  

- Python  
- Pandas  
- NumPy  
- Scikit-learn  

---

## Key Learnings  

- Understanding how multiple features affect predictions  
- Importance of proper data preprocessing  
- Handling categorical variables correctly  
- Building and interpreting regression models  

---

## Next Steps  

- Apply Random Forest Regression
- Compare with other regression models (Multiple Linear Regression, Polynomial Regression, Support Vector Regression (SVR))  
- Improve model performance with feature engineering  

# **Importing The Libaries**

In [46]:
import numpy as np
import matplotlib.pyplot as plt
import pandas as pd

# **Data Cleaning Check Template**
This template is designed to quickly assess the quality of any dataset before building machine learning models or performing analysis.

It provides a structured overview of the dataset by checking for common data issues such as:

- Missing values

- Duplicate rows

- Incorrect data types

- Outliers

- Distribution of numerical features

- Categorical feature consistency

**What This Template Does**

- Displays basic dataset information (shape, data types)

- Identifies missing values and duplicates

- Summarizes numerical and categorical features

- Detects potential outliers using the IQR method

- Highlights columns with low unique values for quick inspection

How to Use

1. Load your dataset using Pandas  

2. Call the function:


In [ ]:
def data_quality_report(df):

    print("DATA QUALITY REPORT")
    
    # Print a separator line for better readability
    
    print("=" * 50)
    print("BASIC INFO")
    print("=" * 50)
    
    # Show general information about the dataset (columns, data types, non-null values)
    print(df.info())
    
    # Show number of rows and columns
    print("\n" + "=" * 50)
    print("SHAPE OF DATA")
    print("=" * 50)
    print(f"Rows: {df.shape[0]}, Columns: {df.shape[1]}")
    
    # Check for missing (null) values in each column
    print("\n" + "=" * 50)
    print("MISSING VALUES")
    print("=" * 50)
    missing = df.isnull().sum()
    
    # Display only columns that have missing values
    print(missing[missing > 0])
    
    # Check for duplicate rows
    print("\n" + "=" * 50)
    print("DUPLICATES")
    print("=" * 50)
    print(f"Duplicate rows: {df.duplicated().sum()}")
    
    # Display data types of each column
    print("\n" + "=" * 50)
    print("DATA TYPES")
    print("=" * 50)
    print(df.dtypes)
    
    # Summary statistics for numerical columns (mean, std, min, max, etc.)
    print("\n" + "=" * 50)
    print("NUMERICAL SUMMARY")
    print("=" * 50)
    print(df.describe())
    
    # Summary for categorical (object) columns
    print("\n" + "=" * 50)
    print("CATEGORICAL SUMMARY")
    print("=" * 50)
    print(df.describe(include=['object']))
    
    # Show unique values for columns with low number of distinct values
    # Useful for detecting categories, errors, or inconsistencies
    print("\n" + "=" * 50)
    print("UNIQUE VALUES (LOW CARDINALITY)")
    print("=" * 50)
    for col in df.columns:
        if df[col].nunique() < 10:  # Only show columns with few unique values
            print(f"{col}: {df[col].unique()}")
            
    # correlation
    print("\n" + "=" * 50)
    print("CORRELATION MATRIX")
    print("=" * 50)
    print(df.corr(numeric_only=True))
    
    # Detect outliers using the IQR (Interquartile Range) method
    print("\n" + "=" * 50)
    print("OUTLIERS CHECK (IQR METHOD)")
    print("=" * 50)
    
    # Loop through only numerical columns
    for col in df.select_dtypes(include=np.number).columns:
        Q1 = df[col].quantile(0.25)  # 25th percentile
        Q3 = df[col].quantile(0.75)  # 75th percentile
        IQR = Q3 - Q1  # Interquartile range
        
        # Count rows that fall outside the normal range
        outliers = df[(df[col] < Q1 - 1.5 * IQR) | (df[col] > Q3 + 1.5 * IQR)]
        print(f"{col}: {len(outliers)} outliers")

# **Load dataset**
Apply Data Cleaning Check Template

In [48]:
dataset = pd.read_csv("/kaggle/input/datasets/aiwithcagri/multiple-linear-regression-tesla-model-y-price/teslay.csv")
data_quality_report(dataset)

DATA QUALITY REPORT
BASIC INFO
<class 'pandas.core.frame.DataFrame'>
RangeIndex: 255 entries, 0 to 254
Data columns (total 4 columns):
 #   Column  Non-Null Count  Dtype  
---  ------  --------------  -----  
 0   year    255 non-null    int64  
 1   km      255 non-null    float64
 2   color   255 non-null    object 
 3   price   255 non-null    int64  
dtypes: float64(1), int64(2), object(1)
memory usage: 8.1+ KB
None

SHAPE OF DATA
Rows: 255, Columns: 4

MISSING VALUES
Series([], dtype: int64)

DUPLICATES
Duplicate rows: 4

DATA TYPES
year       int64
km       float64
color     object
price      int64
dtype: object

NUMERICAL SUMMARY
              year             km         price
count   255.000000     255.000000  2.550000e+02
mean   2024.207843   30384.172549  2.247061e+06
std       0.406562   15563.810709  1.517529e+05
min    2024.000000    6001.000000  1.825000e+06
25%    2024.000000   19880.000000  2.160000e+06
50%    2024.000000   29000.000000  2.239750e+06
75%    2024.000000 

# **Understaning the outliers** 

### km: 6 outliers

In [49]:
# Calculate fences for Price
Q1 = dataset['km'].quantile(0.25)
Q3 = dataset['km'].quantile(0.75)
IQR = Q3 - Q1

lower_fence = Q1 - 1.5 * IQR
upper_fence = Q3 + 1.5 * IQR

# Filter the dataframe to show ONLY the outliers
km_outliers = dataset[(dataset['km'] < lower_fence) | (dataset['km'] > upper_fence)]

# Display the rows
print(km_outliers)

     year        km  color    price
0    2025   69000.0  White  2095000
23   2024   73000.0   Blue  2140000
96   2024   75000.0  White  2010000
145  2024  110000.0  White  2040000
210  2024   96000.0  Black  2049999
229  2024   68200.0  White  2100000


# **Cleaning the data** 

### Remove km outliers (6) & remove the duplication (4)

In [50]:
# Remove rows that exist in km_outliers from the original dataset
dataset_clean = dataset.drop(km_outliers.index).copy()

# Step 2: Remove duplicates FROM the cleaned dataset
dataset_clean = dataset_clean.drop_duplicates().copy()

# Final result
print(dataset_clean)

     year       km  color    price
1    2024  28500.0  Black  2280000
2    2024  50000.0  White  2090000
3    2024  42000.0  White  2180000
4    2024  15000.0  White  2300000
5    2024  36326.0  White  2155000
..    ...      ...    ...      ...
250  2025  24000.0  Black  2390000
251  2024  29000.0  White  2185000
252  2025  19800.0  White  2305000
253  2024  16000.0  White  2250000
254  2024  34000.0  White  2310000

[245 rows x 4 columns]


# **Select Independent Variable (X) and Dependent Variable (y)**

📌 Note: The concepts of Independent and Dependent variables were discussed earlier.

If you would like to review or gain a clearer understanding, please refer to this section: [Go to Independent variable VS Dependent variable Section](https://github.com/Hazem1695/ml-concept-briefs)

In [52]:
# Separate independent variable(features (X)) by selecting all rows and all columns except the last one
X = dataset_clean.iloc[:, :-1].values
# Separate the dependent variable (Target(y)) by selecting the last column
y = dataset_clean.iloc[:, -1].values

# **Encoding categorical data**

## Encoding the Indpendent Variable Using One Hot Encoding

📌 Note: I explained One-Hot Encoding earlier.

If you missed it or want a deeper understanding of how it works, you can find it here: [Go to One-Hot Encoding Section](https://github.com/Hazem1695/ml-concept-briefs)

In [53]:
from sklearn.compose import ColumnTransformer
from sklearn.preprocessing import OneHotEncoder
# One-hot encode the 'color' column (index 2) while keeping other features
# Initialize ColumnTransformer to handle categorical data
# [2] targets the 'color' column (index 2)
ct = ColumnTransformer(transformers = [('encoder', OneHotEncoder(sparse_output=False),[2])], remainder = 'passthrough')
# Execute transformation and convert to NumPy ndarray for model training
X = np.array(ct.fit_transform(X))

# **Splitting the dataset into the Train set and Test set**

In [54]:
from sklearn.model_selection import train_test_split
X_train, X_test, y_train , y_test = train_test_split(X, y , test_size = 0.2, random_state = 0)

In [55]:
print (X_test)

[[0.0 0.0 0.0 0.0 0.0 0.0 0.0 0.0 1.0 2024 25000.0]
 [0.0 0.0 0.0 0.0 0.0 0.0 0.0 0.0 1.0 2024 22700.0]
 [0.0 0.0 0.0 0.0 0.0 0.0 0.0 0.0 1.0 2024 10124.0]
 [0.0 0.0 0.0 0.0 0.0 0.0 0.0 0.0 1.0 2024 33000.0]
 [0.0 0.0 0.0 0.0 0.0 0.0 0.0 0.0 1.0 2024 66000.0]
 [0.0 0.0 0.0 0.0 0.0 0.0 0.0 0.0 1.0 2024 41300.0]
 [0.0 0.0 0.0 0.0 0.0 0.0 0.0 0.0 1.0 2024 7650.0]
 [1.0 0.0 0.0 0.0 0.0 0.0 0.0 0.0 0.0 2024 6001.0]
 [0.0 0.0 0.0 0.0 0.0 0.0 0.0 0.0 1.0 2024 37000.0]
 [0.0 0.0 0.0 0.0 0.0 0.0 0.0 0.0 1.0 2024 44000.0]
 [0.0 0.0 1.0 0.0 0.0 0.0 0.0 0.0 0.0 2024 31000.0]
 [1.0 0.0 0.0 0.0 0.0 0.0 0.0 0.0 0.0 2025 19200.0]
 [0.0 0.0 0.0 0.0 0.0 0.0 0.0 0.0 1.0 2024 6001.0]
 [0.0 0.0 0.0 0.0 0.0 0.0 0.0 0.0 1.0 2025 19200.0]
 [1.0 0.0 0.0 0.0 0.0 0.0 0.0 0.0 0.0 2024 52531.0]
 [0.0 0.0 0.0 0.0 0.0 0.0 0.0 0.0 1.0 2025 41200.0]
 [0.0 0.0 0.0 0.0 0.0 0.0 0.0 0.0 1.0 2024 29000.0]
 [1.0 0.0 0.0 0.0 0.0 0.0 0.0 0.0 0.0 2024 25000.0]
 [1.0 0.0 0.0 0.0 0.0 0.0 0.0 0.0 0.0 2025 41500.0]
 [0.0 1.0 0.0 0

In [56]:
print(f"Rows: {X_test.shape[0]}, Columns: {X_test.shape[1]}")

Rows: 49, Columns: 11


# **Decision Tree Regression Model**

## Training the Building Decision Tree Regression model on the Training set

In [57]:
from sklearn.tree import DecisionTreeRegressor
regressor = DecisionTreeRegressor(max_depth=3, min_samples_split=5, min_samples_leaf=3, random_state = 0)
regressor.fit(X_train, y_train)

DecisionTreeRegressor(max_depth=3, min_samples_leaf=3, min_samples_split=5,
                      random_state=0)

1- `max_depth`: The maximum number of levels in the tree.

- Depth = how many splits from root → leaf
- A deeper tree = more complex model

**Effect:**

- Small depth → underfitting (too simple)
- Large depth → overfitting (too complex)

---

2- `min_samples_split`: The minimum number of samples required to split a node.

- The model will NOT split a node unless it has at least this many samples.

**Effect:**

- Small value → more splits → complex tree
- Large value → fewer splits → simpler tree

---

3- `min_samples_leaf`: The minimum number of samples allowed in a leaf (final node)

- After splitting, each leaf must have at least this many samples.

**Effect:**

- Small value → overfitting risk
- Larger value → smoother predictions


# **Preticting the Test set results**

In [58]:
y_pred = regressor.predict(X_test)
np.set_printoptions(precision=2)
print(np.concatenate((y_pred.reshape(len(y_pred),1), y_test.reshape(len(y_test),1)),1))

[[2224812.35 2185000.  ]
 [2224812.35 2250000.  ]
 [2298326.39 2275000.  ]
 [2224812.35 2249990.  ]
 [2117613.33 2089000.  ]
 [2224812.35 2199000.  ]
 [2298326.39 2350000.  ]
 [2520000.   2425000.  ]
 [2224812.35 2100000.  ]
 [2117613.33 2050000.  ]
 [2224812.35 2320000.  ]
 [2388809.52 2350000.  ]
 [2520000.   2490000.  ]
 [2388809.52 2365000.  ]
 [2206250.   2125000.  ]
 [2297861.76 2400000.  ]
 [2224812.35 2265555.  ]
 [2224812.35 2340000.  ]
 [2297861.76 2180000.  ]
 [2298326.39 2350000.  ]
 [2297861.76 2165000.  ]
 [2206250.   2189000.  ]
 [2224812.35 2185000.  ]
 [2297861.76 2215000.  ]
 [2224812.35 2199000.  ]
 [2224812.35 2240000.  ]
 [2298326.39 2199000.  ]
 [2224812.35 2075000.  ]
 [2520000.   2500000.  ]
 [2388809.52 2285000.  ]
 [2224812.35 1990000.  ]
 [2224812.35 2130000.  ]
 [2298326.39 2300000.  ]
 [2224812.35 2280000.  ]
 [2297861.76 2165000.  ]
 [2224812.35 1950000.  ]
 [2298326.39 2490000.  ]
 [2224812.35 2269900.  ]
 [2117613.33 2195000.  ]
 [2297861.76 2240000.  ]


# **Evaluating the Model Performance**

In [59]:
from sklearn.metrics import r2_score
r2_score(y_test, y_pred)

0.49463534015325283

# **Important Insight**

1. The initial Decision Tree model produced a negative R² score (-0.35), indicating a complete failure to generalize beyond training data.

2. After applying regularization constraints (max_depth=3, min_samples_split=5, min_samples_leaf=3), performance improved significantly to R² = 0.49.

3. This confirms that the issue was not the model itself, but uncontrolled model complexity leading to overfitting.

4. The improvement demonstrates that capacity control is critical, especially when working with small datasets (245 samples).
5. In comparison, SVR achieved the highest performance (R² = 0.53) when preprocessing was applied selectively, reinforcing that both model choice and preprocessing strategy directly impact performance.

---

# **Conclusion**

The experiments show that model performance is highly sensitive to both preprocessing strategy and model complexity.

The Decision Tree model initially failed due to overfitting, as indicated by a negative R² score (-0.35). By introducing constraints to control model complexity, performance improved significantly to 0.49, confirming that generalization depends on proper regularization rather than model selection alone.

Similarly, SVR performance varied substantially depending on preprocessing choices. Scaling all features led to poor results (0.23), while selectively scaling only high-magnitude numerical features improved performance to 0.53, the best result among all models.

These results highlight that effective machine learning is driven by informed design decisions, including selective preprocessing and controlled model capacity, rather than applying standard techniques without validation.

---

# **Model Comparison**

| Model                | Configuration                          | R² Score |
|---------------------|----------------------------------------|---------|
| Linear Regression   | No Scaling                             | 0.42    |
| Linear Regression   | With Scaling                           | 0.41    |
| Polynomial Regression | With Scaling                         | 0.48    |
| SVR                | Scale All Features                      | 0.23    |
| SVR                | Scale Numerical Features (Year, KM)     | 0.53    |
| Decision Tree      | Default (Unconstrained)                 | -0.35   |
| Decision Tree      | Regularized (Controlled Depth)          | 0.49    |